Lib/multiprocessing/shared_memory.py

```diff
@@ -115,57 +113,146 @@ def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
+
+
+def _make_filename():
+    """Create a random filename for the shared memory object.
+    """
+    # number of random bytes to use for name
+    nbytes = (_SHM_SAFE_NAME_LENGTH - len(_SHM_NAME_PREFIX)) // 2
+    assert nbytes >= 2, '_SHM_NAME_PREFIX too long'
+    name = _SHM_NAME_PREFIX + secrets.token_hex(nbytes)
+    assert len(name) <= _SHM_SAFE_NAME_LENGTH
+    return name
+
+
+class PosixSharedMemory:
+
+    # defaults so close() and unlink() can run without errors
+    fd = -1
+    name = None
+    _mmap = None
+    _buf = None
 
     def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
         if name and (flags is None):
-            _PosixSharedMemory.__init__(self, name, mode=mode)
+            flags = 0
         else:
-            if name is None:
-                name = f'psm_{os.getpid()}_{random.randrange(100000)}'
             flags = O_CREX if flags is None else flags
-            _PosixSharedMemory.__init__(
-                self,
-                name,
-                flags=flags,
-                mode=mode,
-                size=size,
-                read_only=read_only
-            )
-
+        if flags & O_EXCL and not flags & O_CREAT:
+            raise ValueError("O_EXCL must be combined with O_CREAT")
+        if name is None and not flags & O_EXCL:
+            raise ValueError("'name' can only be None if O_EXCL is set")
+        flags |= os.O_RDONLY if read_only else os.O_RDWR
+        self.flags = flags
+        self.mode = mode
+        if not size >= 0:
+            raise ValueError("'size' must be a positive integer")
+        if name is None:
+            self._open_retry()
+        else:
+            self.name = name
+            self.fd = _posixshmem.shm_open(name, flags, mode=mode)
+        if size:
+            try:
+                os.ftruncate(self.fd, size)
+            except OSError:
+                self.unlink()
+                raise
         self._mmap = mmap.mmap(self.fd, self.size)
-        self.buf = memoryview(self._mmap)
+        self._buf = memoryview(self._mmap)
+
+    @property
+    def size(self):
+        "Size in bytes."
+        if self.fd >= 0:
+            return os.fstat(self.fd).st_size
+        else:
+            return 0
+
+    @property
+    def buf(self):
+        "A memoryview of contents of the shared memory block."
+        return self._buf
+
+    def _open_retry(self):
+        # generate a random name, open, retry if it exists
+        while True:
+            name = _make_filename()
+            try:
+                self.fd = _posixshmem.shm_open(name, self.flags,
+                                               mode=self.mode)
+            except FileExistsError:
+                continue
+            self.name = name
+            break
 
     def __repr__(self):
         return f'{self.__class__.__name__}({self.name!r}, size={self.size})'
 
+    def unlink(self):
+        if self.name:
+            _posixshmem.shm_unlink(self.name)
```

Comment:
shouldn't `unlink` also call `close`?

Reply:
In the same sense that ``close()`` should not call ``unlink()``, the reverse should also be true to maintain a clean separation of duties.  This keeps things clear in situations like the following: 
 if the coordinating process is done accessing the shared memory block, it should be permitted to call ``close`` at that time and wait to call ``unlink`` once it knows all other processes are similarly done; in this case we would not want ``unlink`` to call ``close`` a second time.  After all, if ``unlink`` did call ``close``, it would only be able to do so in one process but other processes should still call ``close``.  Best to keep them separated.  There is a reason why this is the design for the shared memory C APIs that we use on various operating systems.

Reply:
Clear enough. Thanks.

Reply:
On a second thought... if `unlink()` is called, isn't the instance supposed to be unusable? AFAICT you can still play with the memoryview object but only locally (no data is shared). Also, I wonder whether it makes sense to protect `buf` attribute so that it raises an exception if used after `close()` and `unlink()`.

Reply:
If one process calls ``unlink()``, we might protect ``buf`` and even do all the other things that ``close()`` would do for that one process but all other processes still need to call ``close()`` themselves.  I think the clean separation of duties is still the right answer here.

The docs currently say that if ``unlink()`` is called, any continued use is unsupported.  The exact behavior of ``unlink`` with respect to when the shared memory block truly is destroyed is not something we can ultimately control or enforce.  Users must call ``close()`` and ``unlink()`` to ensure proper cleanup.  If users can not follow this directive, I doubt we could ever truly protect them from themselves.